implement image pull secrets

deads2k opened the pull request:
Adding image pull secrets in etcd to handle registries that require authentication.

TODO
- [x] add dockercfg secrets
- [x] add command to make it easy to create secrets
- [x] update the docker keyring to return a list of credentials to try
- [x] update `.Pull` to handle config based versus host based credentials
- [x] wire up a pull secret list

@liggitt @pmorie

deads2k committed:
add dockercfg secret types

deads2k committed:
add helper command for creating dockerconfig secrets

brendandburns commented:
Travis is grumpy on gofmt issues.  Other than that, ping this when it is ready to review.

Thanks!
--brendan

liggitt commented on pkg/api/types.go:
```diff
@@ -1801,6 +1801,15 @@ type SecretType string
+	//
+	// Required fields:
+	// - Secret.Data["dockercfg"] - a serialized ~/.dockercfg file
+	SecretTypeDockerConfig SecretType = "DockerConfig"
```
just occurred to me that we may want to qualify secret types (e.g. prefix with "kubernetes.io/")

liggitt replied:
don't forget `api/v1` :)

deads2k replied:
done

liggitt commented on pkg/kubectl/cmd/secrets/dockercfg.go:
```diff
@@ -0,0 +1,169 @@
+	Out io.Writer
+}
+
+// NewCmdGet creates a command object for the generic "get" action, which
```
copy/paste

liggitt replied:
copy/paste

deads2k replied:
done.  But not twice :)

liggitt commented on pkg/kubectl/cmd/secrets/dockercfg.go:
```diff
@@ -0,0 +1,169 @@
+	}
+	if len(o.Username) == 0 {
+		return errors.New("docker-username must be present")
+	}
```
should we prevent usernames that contain `:`, since they can't be encoded into a basic-auth header?

deads2k replied:
done

liggitt commented on pkg/credentialprovider/keyring.go:
```diff
@@ -73,15 +76,19 @@ func (dk *BasicDockerKeyring) Add(cfg DockerConfig) {
-		if parsed.Path != "/" {
-			dk.creds[parsed.Host+parsed.Path] = creds
-			dk.index = append(dk.index, parsed.Host+parsed.Path)
+		if (len(parsed.Path) > 0) && (parsed.Path != "/") {
```
if the path doesn't start with `/`, we could get conflicts between hosts and host+path combos

deads2k replied:
> if the path doesn't start with /, we could get conflicts between hosts and host+path combos

That is pre-existing.  I had assumed that `Url.Path` would always have a slash.

liggitt commented on pkg/credentialprovider/keyring.go:
```diff
@@ -109,11 +116,12 @@ func isDefaultRegistryMatch(image string) bool {
-	// a predictable ordering
+// Lookup implements the DockerKeyring method for fetching credentials based on image name.
+// Multiple credentials may be returned if there are multiple potentially valid credentials
+// avaiable.  This allows for rotation.
```
typo

liggitt commented on pkg/credentialprovider/keyring.go:
```diff
@@ -121,7 +129,16 @@ func (dk *BasicDockerKeyring) Lookup(image string) (docker.AuthConfiguration, bo
 		}
 
-		return dk.creds[k], true
+		dockerAuthKeyMatches = append(dockerAuthKeyMatches, k)
```
this changes the behavior to include credentials for a parent path when a more specific set of credentials are available... is that what we want?

liggitt replied:
also, could just inline `ret = append(ret, dk.creds[k]...)` here

deads2k replied:
> this changes the behavior to include credentials for a parent path when a more specific set of credentials are available... is that what we want?

Yes.  If some of the credentials are expired, then it makes sense to try the less specific ones after.

deads2k replied:
We always try the most specific first.

liggitt commented on pkg/kubectl/cmd/secrets/dockercfg.go:
```diff
@@ -0,0 +1,169 @@
+
+// NewCmdGet creates a command object for the generic "get" action, which
+// retrieves one or more resources from a server.
+func NewCmdCreateDockerConfigSecrets(name, fullName string, f *cmdutil.Factory, out io.Writer) *cobra.Command {
```
singular?

deads2k replied:
done

deads2k commented:
@brendandburns completed and rebased.  rkt only handles a single credential for now.

pmorie commented on pkg/kubelet/rkt/rkt_linux.go:
```diff
@@ -742,7 +748,7 @@ func (r *runtime) writeDockerAuthConfig(image string, creds docker.AuthConfigura
 // https://github.com/GoogleCloudPlatform/kubernetes/issues/7203
 //
-func (r *runtime) PullImage(image kubecontainer.ImageSpec, _ []api.Secret) error {
+func (r *runtime) PullImage(image kubecontainer.ImageSpec, pullSecrets []api.Secret) error {
```
Are you planning on making this func use pullSecrets in this PR?

deads2k replied:
> Are you planning on making this func use pullSecrets in this PR?

rebase movement.  Refactored to use it more properly.  It's still only first hit wins (as it was before).

yifan-gu commented on pkg/api/types.go:
```diff
@@ -820,6 +820,10 @@ type PodSpec struct {
+	// PullSecrets is an optional list of secret names to use for pulling any of the images used by this PodSpec.
+	// If specified, these secrets will be passed to individual puller implementations for them to use.  For example,
+	// in the case of docker, only DockerConfig type secrets are honored.
+	PullSecrets []string `json:"pullSecrets,omitempty"`
```
nit: PullSecretNames? The PullSecrets is a little bit misleading to me. Though I know we won't be lucky enough to get real secrets in plain json...

deads2k replied:
> nit: PullSecretNames? The PullSecrets is a little bit misleading to me. Though I know we won't be lucky enough to get real secrets in plain json...

done.

deads2k commented:
is `pkg/kubelet/rkt` intentionally disconnected from the rest of the packages?  It doesn't build with `make` or `hack/build-go.sh`, but `go install ./...` uses it.

brendandburns commented on pkg/api/types.go:
```diff
@@ -1803,6 +1811,15 @@ type SecretType string
+	// SecretTypeDockerConfig contains a dockercfg file that follows the same format rules as ~/.dockercfg
+	//
+	// Required fields:
+	// - Secret.Data["dockercfg"] - a serialized ~/.dockercfg file
```
it might clarify things if you put 'DockerConfigKey' instead of "dockercfg" as the key, since it will establish the linkage between the two constants.

deads2k replied:
> it might clarify things if you put 'DockerConfigKey' instead of "dockercfg" as the key, since it will establish the linkage between the two constants.

I was trying to keep the name as close as possible to do the file that docker uses by default: `~/.dockercfg`.  If you're bundling a secret automatically via a helper command, stripping the leading '.' to make a dns1123 compatible name is natural.  That means that this secret could be made via: `bundle-secret ~/.dockercfg`.

liggitt replied:
+1

liggitt replied:
I would actually change the secret type to `kubernetes.io/dockercfg` :)

brendandburns commented:
This looks reasonable to me, but it actually feels like 3 PRs:
  1. Add DockerCfg secrets to the Pod API
  2. Implement support for DockerCfg secrets in kubelet 
  3. Add secrets support to kubectl
  4. Add dockercfg support to secrets to kubectl

Can you split this into those PRs?  it will make it both easier to review and to merge (or even rollback if we need to)

I could see the case for merging 1 and 2 if you like...

Thanks!
--brendan

deads2k commented:
> This looks reasonable to me, but it actually feels like 4 PRs:

I'm splitting it apart now.  I'll take you on the new pulls.

deads2k committed:
add pull secret references to pods

deads2k commented:
separate pulls created and linked back to this one.

Updated names for @liggitt

deads2k commented:
rebased.

@smarterclayton you want to take a look?  This PR contains all the commits, the references PRs are the split commits as per https://github.com/GoogleCloudPlatform/kubernetes/pull/7894#issuecomment-100290585

smarterclayton commented:
Looking

bgrant0607 commented:
Starting with a basic question: Is .dockercfg even the right file to use any longer, or should we be targeting .docker/config.json?

https://github.com/docker/docker/blob/master/cliconfig/config.go#L18

Unfortunately, I don't have time to look at these PRs today. I'll look at the whole batch on Tuesday or Wednesday.

smarterclayton commented:
Definitely reduces the clarity.  Shame.  dockerconfig.json feels more correct if that's the case.  .dockerconfig.json seems wrong, and .dockercfg would be out of date.

----- Original Message -----
> Starting with a basic question: Is .dockercfg even the right file to use any
> longer, or should we be targeting .docker/config.json?
> 
> https://github.com/docker/docker/blob/master/cliconfig/config.go#L18
> 
> Unfortunately, I don't have time to look at these PRs today. I'll look at the
> whole batch on Tuesday or Wednesday.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/7894#issuecomment-100997903

deads2k committed:
allow leading dots in secret keys

erictune commented on pkg/api/types.go:
```diff
@@ -822,6 +822,10 @@ type PodSpec struct {
+	// PullSecretNames is an optional list of secret names to use for pulling any of the images used by this PodSpec.
+	// If specified, these secrets will be passed to individual puller implementations for them to use.  For example,
+	// in the case of docker, only DockerConfig type secrets are honored.
+	PullSecretNames []string `json:"pullSecretNames,omitempty" description:"list of secret names available for pulling the container images"`
```
Just so I understand: these secrets are not visible to any container in the pod, right?

liggitt replied:
right, not mounted into the container by virtue of being in this list. this is for the Kubelet (or other puller's) purposes only

erictune replied:
I don't think a list of strings works well with patching.
See https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/api-conventions.md and search for "strategic merge patch".

This should to be fixed.

erictune replied:
What is the planned behavior for the admission controller if the pod is POSTed with some values in this array?  Is it to only use the user-provided values, or to merge?  I think the preferred behavior would be to not set any values if the user has set any.

In that case, this would ideally be a pointer to a slice rather than a slice.  That way the admission controller can detect the difference between unset and explicitly empty list.

deads2k replied:
> This should to be fixed.

Just choosing a patchStrategy to use as a default or creating a container object with a name?

liggitt replied:
> I don't think a list of strings works well with patching

`ValidatePodUpdate` would not currently allow changing this... I think maybe it should, since it relates to container images, which are mutable.

> Just choosing a patchStrategy to use as a default or creating a container object with a name?

patchStrategy on a scalar list just merges, which means you can't patch to remove an item. That's why the `Secrets` field for a `ServiceAccount` is currently `[]api.ObjectReference` scrubbed of everything except `Name`

deads2k replied:
@smarterclayton brought up the question of whether there was value in having this on a pod at all.  See https://github.com/deads2k/kubernetes/commit/e36818ac2ec1af0c5dd24925c5f4b0f90c778ef1 .  I don't feel very strongly.  I thought having it on a pod would be convenient (and it certainly was for testing), but I hadn't really considered whether it was essential.  We could force someone to simply use a service account.

smarterclayton replied:
----- Original Message -----
> > @@ -822,6 +822,10 @@ type PodSpec struct {
> >  	// used must be specified.
> >  	// Optional: Default to false.
> >  	HostNetwork bool `json:"hostNetwork,omitempty"`
> > +	// PullSecretNames is an optional list of secret names to use for pulling
> > any of the images used by this PodSpec.
> > +	// If specified, these secrets will be passed to individual puller
> > implementations for them to use.  For example,
> > +	// in the case of docker, only DockerConfig type secrets are honored.
> > +	PullSecretNames []string `json:"pullSecretNames,omitempty"
> > description:"list of secret names available for pulling the container
> > images"`
> 
> What is the planned behavior for the admission controller if the pod is
> POSTed with some values in this array?  Is it to only use the user-provided
> values, or to merge?  I think the preferred behavior would be to not set any
> values if the user has set any.
> 
> In that case, this would ideally be a pointer to a slice rather than a slice.
> That way the admission controller can detect the difference between unset
> and explicitly empty list.

JSON and our conversion framework already preserves nil slice vs empty slice (or did).  So we don't have to have pointer to slice.

erictune replied:
@bgrant0607 can you comment here?

liggitt replied:
The ServiceAccount PR  lets you decide to not use service accounts (by omitting the admission plugin) and just manage pods/secrets yourself. API tokens, pull secrets, and security context restrictions seem like things that would be valuable to group under a service account, but I'm not sure about requiring service accounts to use those things.

erictune commented:
How does this interact with the keyring that is present on the node?
Does it get merged with that one, or replace it?

erictune commented:
looks good overall.

deads2k commented:
> How does this interact with the keyring that is present on the node?
> Does it get merged with that one, or replace it?

Specifying pull secrets on a pod or via service account (once it merges), will replace the keyring from a node.  That's needed to provide consistent pull behavior across nodes.

erictune commented:
When you say "consistent", is the concern that a cluster admin might
mistakenly put different keyrings on different nodes?  Seems like operator
error to me, and not a reason to not support merging.

If we are going to have "replace" behavior, rather than "merge" then I
think we should just deprecate node-based keyrings.

On Mon, May 11, 2015 at 12:46 PM, David Eads <notifications@github.com>
wrote:

> How does this interact with the keyring that is present on the node?
> Does it get merged with that one, or replace it?
>
> Specifying pull secrets on a pod or via service account (once it merges),
> will replace the keyring from a node. That's needed to provide consistent
> pull behavior across nodes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7894#issuecomment-101028565>
> .
>

erictune commented:
Having a PullSecretNames allows kubelet to not have to understand ServiceAccounts.  So, I like having it.

smarterclayton commented:
SGTM 

> On May 12, 2015, at 1:09 AM, Eric Tune <notifications@github.com> wrote:
> 
> Having a PullSecretNames allows kubelet to not have to understand ServiceAccounts. So, I like having it.
> 
> —
> Reply to this email directly or view it on GitHub.
>

deads2k commented:
1. updated secret key to be ".dockercfg".
 1. updated `PodSpec` to use `[]ObjectReference`.  Changed named to `PullSecrets`, since an `ObjectReference` obviously refers to another object.

deads2k commented:
> When you say "consistent", is the concern that a cluster admin might
> mistakenly put different keyrings on different nodes?  Seems like operator
> error to me, and not a reason to not support merging.
> 
> If we are going to have "replace" behavior, rather than "merge" then I
> think we should just deprecate node-based keyrings.

I prefer "replace" behavior, because I think that it is easier to reason about and makes it possible to reason about the success or failure of an image pull based on the information in etcd, as opposed to the invisible (to API users) configuration of a node.

liggitt commented:
> I prefer "replace" behavior

I find it counter-intuitive that by adding pull credentials to a serviceaccount/pod, I could actually lose permissions I would have otherwise had by default by virtue of node-wide credentials.

deads2k commented:
> I find it counter-intuitive that by adding pull credentials to a serviceaccount/pod, I could actually lose permissions I would have otherwise had by default by virtue of node-wide credentials.

Do you like being able to look at a pod and unambiguously know whether the pull will fail without knowledge of the configuration on the node?  With a merge, you could unambiguously know success because you could add a pull secret, but you could not unambiguously know failure.  It might work, but not because your pull secret is correct and you won't know the difference until the node's config rots.

liggitt commented:
> Do you like being able to look at a pod and unambiguously know whether the pull will fail without knowledge of the configuration on the node?

I dunno... I like my pods working... I don't really care why they work

liggitt commented:
To be clear, I would give priority to the pull secrets specified explicitly for the pod... the node's pull credentials should go at the end of the list to try

deads2k commented:
> the node's pull credentials should go at the end of the list to try

No.  Specificity should win.  Perhaps after specificity, there's an argument for node config last, but I don't really think that it matters.  I'd still rather see replace behavior, so that you can fully reason about results without access to the nodes.

liggitt commented:
> No. Specificity should win.

Sorry, I meant within a given `path -> [secrets]` list, the node's should go at the end

deads2k commented:
The individual pulls (minus the command) are in.  Closing this.
